### PR TITLE
`requestPerItem` more complex requests with params, async & params as array

### PR DESCRIPTION
**Pull request by fratzinger**
### TL;DR:
**Features:**
- wait for async `params` function
- can pass `params` as array -> mostly meant for other libraries (as `feathers-graph-populate`) (I will come to this on `feathers-graph-populate)`
- access `this` keyword in `params` function for individual items
  - leave `keyHere` and `keyThere` to perform a custom request per item
  - sets `requestPerItem` for these includes automatically -> for `requestPerItem: true` includes: run requests per item
- added jsdoc comments for better IDE support
- support for `$skip` and `$limit`
- support for `$select` which is missing `${keyThere}`
- allow throwing populates to pass
- added tests for these changes of course

**Changed:**
- created `utils.js` with assert to clean up the hook
- moved params creation to `utils.js`
- use async/await
- tests: 
  - removed `done` callback in favor of async/await.
  - refactor for `before` and `after` to clean it up

### ❗  Whats the problem?
**1. complex requests:** Simple `keyHere` and `keyThere` relations are the most common use case. This library has a beautiful approach for handling these. I decided using this along `feathers-graph-populate` because of its simpleness and high readability. But it lacks if I have some more complex relations (m-n, or tables sharing multiple keys). Because I want to define these also in `feathers-graph-populate` I made the changes in this PR.
**2. support `$skip`/`$limit`:** `feathers-shallow-populate` has no support for filters `$skip` and `$limit` since it combines multiple requests into one.
**3. `$select` needs `keyThere`:** When using the `$select` filter, it needs at least to have `$select: [include.keyHere]` which isn't intuitive if you don't know the library.
**4. does not catch:** A real world application needs to pass permissions to the populates hook. Permissions then use to throw an error, if the user is not allowed to do the population. I think there a apps which need to respond with the result but without the populated record. Therefore it needs to catch errors.

### ✔️ How is it solved?
This PR is not a breaking change. It won't affect others.
**1. complex requests:**
For more complex relations we can drop the `keyHere` and `keyThere` attribute and the params-function need to access the individual item. The `params` attribute therefore gets access to `this` keyword being the underlying item in `data`/`result`. Since even more complex requests need to call other services first, it's possible to define a async `params`-function. `feathers-shallow-populate` waits for them to return before it continues. Example:

```
{
  service: 'orgs',
  nameAs: 'org',
  asArray: false,
  params: async function (params, context) {
    if (!this.userId) return undefined;
    const user = await context.app.service('users').get(this.userId)
    if (!user) return undefined;
    return { query: { id: user.orgId } }
  }
},
```

if the `params` function returns undefined, it will be skipped for the particular item

**2. support `$skip`/`${limit`}:**
It just works as intended now. It removes these filters before the request to the server and regards them afterwards

**3. `$select` needs `keyThere`:**
It just works as intended now. If `$select` is defined and hasn't the `${include.keyThere}` before performing the request, it gets added before automatically.

**4. does not catch:**
That became an option to not break existing applications. You can set it either with the hooks options `makePopulate({ catchOnError: true, include: ...)` or more dynamically on the particular include: `{ service: 'posts', catchOnError: true }`. It's `false` by default.

### Maybe something for later:

- `keyHere` and ``keyThere` as arrays -> matches all keys
- define `service`, `nameAs`, `keyThere`, `keyHere`, `asArray` as function with `(params, context): (string|Promise<string>)`

I'll start now implementing this into `feathers-graph-populate`. I'll let you know when both align @marshallswain :)

Closing #18

### CODE REVIEW

**Comment by fratzinger** [comment]
I think it's ready.
@marshallswain what do you think?

**Comment by marshallswain** [comment]
Oh man!  This is a beautiful thing to see when returning to OS work!  Thanks @fratzinger!

**Comment by marshallswain** [comment]
Released in `feathers-shallow-populate@2.2.0`.  Thanks!

**Comment by fratzinger** [comment]
Great :) Thanks as well :)
